File: 18.HashtableAndOperations/HashTableandOperations.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None


class HashTable:
    def __init__(self):
        self.size = 10
        self.table = [None] * self.size

    def hashFunc(self, key):
        return key % self.size

    def insert(self, key, value):
        index = self.hashFunc(key)
        newNode = Node(key, value)

        newNode.next = self.table[index]   # insert at front
        self.table[index] = newNode

        print(f"Inserted: ({key},{value})")

    def deleteKey(self, key):
        index = self.hashFunc(key)
        curr = self.table[index]
        prev = None

        while curr:
            if curr.key == key:
                if prev:
                    prev.next = curr.next
                else:
                    self.table[index] = curr.next
                print(f"Key {key} deleted successfully.")
                return
            prev = curr
            curr = curr.next

        print(f"Key {key} not found.")

    def search(self, key):
        index = self.hashFunc(key)
        curr = self.table[index]

        while curr:
            if curr.key == key:
                print(f"{key} -> {curr.value}")
                return
            curr = curr.next

        print(f"Key {key} not found.")

    def display(self):
        for i in range(self.size):
            curr = self.table[i]
            if curr:
                print(f"Index {i}:", end=" ")
                while curr:
                    print(f"({curr.key},{curr.value})", end=" ")
                    curr = curr.next
                print()


ht = HashTable()

while True:
    try:
        choice = int(input())
    except:
        break

    if choice == 1:
        key, value = map(int, input().split())
        ht.insert(key, value)

    elif choice == 2:
        key = int(input())
        ht.deleteKey(key)

    elif choice == 3:
        key = int(input())
        ht.search(key)

    elif choice == 4:
        ht.display()

    elif choice == 5:
        print("Exit")
        break

File: 18.HashtableAndOperations/HashTableandOperations.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 10

typedef struct Node {
    int key, value;
    struct Node* next;
} Node;

Node* hashTable[SIZE] = {NULL};

int hash(int key) {
    return key % SIZE;
}

void insert(int key, int value) {
    int index = hash(key);
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->key = key;
    newNode->value = value;
    newNode->next = hashTable[index];  // insert at front
    hashTable[index] = newNode;
    printf("Inserted: (%d,%d)\n", key, value);
}

void deleteKey(int key) {
    int index = hash(key);
    Node* curr = hashTable[index];
    Node* prev = NULL;
    while (curr) {
        if (curr->key == key) {
            if (prev) prev->next = curr->next;
            else hashTable[index] = curr->next;
            free(curr);
            printf("Key %d deleted successfully.\n", key);
            return;
        }
        prev = curr;
        curr = curr->next;
    }
    printf("Key %d not found.\n", key);
}

void search(int key) {
    int index = hash(key);
    Node* curr = hashTable[index];
    while (curr) {
        if (curr->key == key) {
            printf("%d -> %d\n", key, curr->value);
            return;
        }
        curr = curr->next;
    }
    printf("Key %d not found.\n", key);
}

void display() {
    for (int i = 0; i < SIZE; i++) {
        if (hashTable[i]) {
            printf("Index %d: ", i);
            Node* curr = hashTable[i];
            while (curr) {
                printf("(%d,%d) ", curr->key, curr->value);
                curr = curr->next;
            }
            printf("\n");
        }
    }
}

int main() {
    int choice, key, value;
    while (1) {
        if (scanf("%d", &choice) != 1) break;
        switch (choice) {
            case 1:
                scanf("%d %d", &key, &value);
                insert(key, value);
                break;
            case 2:
                scanf("%d", &key);
                deleteKey(key);
                break;
            case 3:
                scanf("%d", &key);
                search(key);
                break;
            case 4:
                display();
                break;
            case 5:
                printf("Exit\n");
                return 0;
        }
    }
    return 0;
}
